module/zfs/arc.c

```diff
@@ -3512,22 +3512,44 @@ arc_hdr_realloc_crypt(arc_buf_hdr_t *hdr, boolean_t need_crypt)
 	arc_buf_hdr_t *nhdr;
 	arc_buf_t *buf;
 	kmem_cache_t *ncache, *ocache;
+	unsigned nsize, osize;
 
 	ASSERT(HDR_HAS_L1HDR(hdr));
 	ASSERT3U(!!HDR_PROTECTED(hdr), !=, need_crypt);
 	ASSERT3P(hdr->b_l1hdr.b_state, ==, arc_anon);
 	ASSERT(!multilist_link_active(&hdr->b_l1hdr.b_arc_node));
+	ASSERT(!list_link_active(&hdr->b_l2hdr.b_l2node));
+	ASSERT3P(hdr->b_hash_next, ==, NULL);
 
 	if (need_crypt) {
 		ncache = hdr_full_crypt_cache;
+		nsize = sizeof (hdr->b_crypt_hdr);
 		ocache = hdr_full_cache;
+		osize = HDR_FULL_SIZE;
 	} else {
 		ncache = hdr_full_cache;
+		nsize = HDR_FULL_SIZE;
 		ocache = hdr_full_crypt_cache;
+		osize = sizeof (hdr->b_crypt_hdr);
 	}
 
 	nhdr = kmem_cache_alloc(ncache, KM_PUSHPAGE);
-	bcopy(hdr, nhdr, HDR_L2ONLY_SIZE);
+
+	/*
+	 * Copy all members that aren't locks or condvars to the new header.
+	 * No lists are pointing to us (as we asserted above), so we don't
+	 * need to worry about the list nodes.
+	 */
```

Comment:
This seems substantially different from arc_hdr_realloc(), which uses bcopy() (and doesn't clear/bzero the old hdr).  Can you explain why this difference is needed?

Reply:
The bcopy doesn't work here because (at least not easily) because there are fields that won't work if they are simply copied (such as the `b_buf` and `b_refcount`). arc_hdr_realloc()` has much less complexity since the only thing it needs to figure out after the copy are the 2 lists that the original header could be on. I could possibly still do this with a bcopy, but I also wasn't completely comfortable bcopying the `b_freeze_lock` and `b_cv`, since that sounds like its asking for trouble.

Reply:
Let me know if you have a better idea for how to solve this problem.